SmartDeviceLink/SDLCarWindow.m

```diff
@@ -0,0 +1,257 @@
+        }
+
+        if (self.isLockScreenMoving) {
+            SDLLogD(@"Paused CarWindow, lock screen moving");
+            return;
+        }
+        
+        self.lastMd5HashTimestamp = displayLink.timestamp;
+        
+        CGRect bounds = self.rootViewController.view.bounds;
+        
+        UIGraphicsBeginImageContextWithOptions(bounds.size, YES, 1.0f);
+        [self.rootViewController.view drawViewHierarchyInRect:bounds afterScreenUpdates:YES];
+        UIImage *screenshot = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        
+        CGImageRef imageRef = screenshot.CGImage;
+        
+        // We use MD5 Hashes to determine if we are sending the same frame over and over. If so, we will only send 30.
+        NSString *currentMd5Hash = [self.class sdl_md5HashForImageRef:imageRef];
+        if ([currentMd5Hash isEqualToString:self.previousMd5Hash]) {
+            if (self.sameFrameCounter == 30) {
+                return;
+            }
+            self.sameFrameCounter++;
+        } else {
+            self.sameFrameCounter = 0;
+        }
+
+        self.previousMd5Hash = currentMd5Hash;
+
+        CVPixelBufferRef pixelBuffer = [self.class sdl_pixelBufferForImageRef:imageRef usingPool:self.streamManager.pixelBufferPool];
+        [self.streamManager sendVideoData:pixelBuffer];
+        CVPixelBufferRelease(pixelBuffer);
+    });
+}
+
+#pragma mark - SDLNavigationLockScreenManager Notifications
+- (void)sdl_lockScreenMoving:(NSNotification *)notification {
+    self.lockScreenMoving = YES;
+}
+
+- (void)sdl_lockScreenStoppedMoving:(NSNotification *)notification {
+    self.lockScreenMoving = NO;
+}
+
+#pragma mark - SDLNavigationLifecycleManager Notifications
+- (void)sdl_didReceiveVideoStreamStarted:(NSNotification *)notification {
+    dispatch_async(dispatch_get_main_queue(), ^{
+        // If the video stream has started, we want to resize the streamingViewController to the size from the RegisterAppInterface
+        self.rootViewController.view.frame = CGRectMake(0, 0, self.streamManager.screenSize.width, self.streamManager.screenSize.height);
+        self.rootViewController.view.bounds = self.rootViewController.view.frame;
+
+        SDLLogD(@"Video stream started, setting CarWindow frame to: %@", NSStringFromCGRect(self.rootViewController.view.bounds));
+        
+        // And reset the frame counter (incase we are coming from a disconnect).
+        self.sameFrameCounter = 0;
+        
+        // And start up the displayLink
+        self.displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(sdl_sendFrame:)];
+        if (SDL_SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"10")) {
+            self.displayLink.preferredFramesPerSecond = self.targetFramerate;
+        } else {
+            self.displayLink.frameInterval = 60/self.targetFramerate;
+        }
+        [self.displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
```

Comment:
Would there be any benefit to putting the displaylink on its own thread, similar to how it is in the iap transport class?

Reply:
No, I don't believe so. The screenshotting must be done on the main thread, and the displaylink is simply calling the method synced with the framerate, so it would likely be detrimental to do so.